File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.cpp
class Solution {
public:
    // store the nxt = nums[curr] to nums[nxt],
    // if nums[nxt] == nxt, return nxt 
    int store(vector<int>& nums, int curr) {
        if (nums[curr] == curr) return curr;
        int nxt = nums[curr];
        nums[curr] = curr;
        return store(nums, nxt);
    }
    
    int findDuplicate(vector<int>& nums) {
        return store(nums, 0);
    }
};

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.py
# Theorm: 
#   For n+1 integer array containing integer in the range [1, n] inclusive, 
#      cannot have every array[n] to hold exactly one `n`. 
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        # Goal: "pigeonhole every integer until encounter a duplicate"
        curr = 0
        nxt = nums[curr]
        while nxt != nums[nxt]:
            curr, nums[nxt] = nums[nxt], nxt
            nxt = curr
            
        return nxt